Sources/Fuzzilli/Core/ProgramBuilder.swift

```diff
@@ -702,18 +702,37 @@ public class ProgramBuilder {
         //   Otherwise copy the whole block including its content
         var requiredInputs = VariableSet()
 
+        // A Set of variables that have yet to be included in the slice
+        var remainingInputs = VariableSet()
+
+        // A stack of contexts that are required by the instruction in the slice
+        var requiredContextStack = [Context.empty]
+
+        // Helper function to add a context to the context stack
+        func addContextRequired(requiredContext: Context) {
+            var currentContext = requiredContextStack.removeLast()
+            currentContext.formUnion(requiredContext)
+            requiredContextStack.append(currentContext)
+        }
+
         // Helper function to add an instruction, or possibly multiple instruction in the case of blocks, to the slice.
         func add(_ instr: Instruction, includeBlockContent: Bool = false) {
             guard !slice.contains(instr.index) else { return }
 
             func internalAdd(_ instr: Instruction) {
+                remainingInputs.subtract(instr.allOutputs)
+
                 requiredInputs.formUnion(instr.inputs)
+                remainingInputs.formUnion(instr.inputs)
+                addContextRequired(requiredContext: instr.op.requiredContext)
```

Comment:
I've added the helper function here. I think from a previous comment you wanted me to add this further below but I'm not sure where you meant that to be?